index.js

```diff
@@ -44,149 +44,162 @@ module.exports = (UserProps) => {
   return {
     postcssPlugin: 'postcss-jit-props',
 
-    async Once (node, {Rule, AtRule}) {
-      let target_selector = ':root'
+    prepare() {
+      const UserPropsCopy = JSON.parse(JSON.stringify(UserProps));
 
-      if (!Object.keys(UserProps).length) {
-        return console.warn('postcss-jit-props: Variable source(s) not passed.')
-      }
-
-      if (UserProps?.files?.length) {
-
-        const files = UserProps?.files
-          .map((file) => glob(file))
-          .reduce((flattenedFileList, files) => flattenedFileList.concat(files), [])
+      return {
+        async Once(node, { result, Rule, AtRule }) {
+          let target_selector = ':root'
 
-        await Promise.all(files.map(async file => {
-          let data = fs.readFileSync(file, 'utf8')
-          let result = await postcss([(function(){})]).process(data, { from: file })
+          if (!Object.keys(UserPropsCopy).length) {
+            return console.warn('postcss-jit-props: Variable source(s) not passed.')
+          }
 
-          result.root.walkDecls(decl => {
-            if (!decl.prop.includes('--')) return
-            UserProps[decl.prop] = decl.value
-          })
+          if (UserPropsCopy?.files?.length) {
+
+            const files = UserPropsCopy?.files
+              .map((file) => glob(file))
+              .reduce((flattenedFileList, files) => flattenedFileList.concat(files), [])
+
+            await Promise.all(files.map(async file => {
+              result.messages.push({
+                type: 'dependency',
+                plugin: 'postcss-jit-props',
+                file: file,
+                parent: node.source?.input?.file,
+              });
+
+              let data = fs.readFileSync(file, 'utf8')
+              let dependencyResult = await postcss([(function () { })]).process(data, { from: file })
```

Comment:
Isn't it a bottleneck? It appears that this will be processed repeatedly for every matching source file. Is it feasible to place it outside of the prepare callback or even the plugin return object? This way, the provided props and state will only be processed once. Thank you.

Reply:
This change doesn't alter that aspect.
It might be a bottleneck but then it already was, this doesn't introduce any regressions.

https://github.com/GoogleChromeLabs/postcss-jit-props/blob/main/index.js#L47-L65

Reply:
It is however easy to add some cache here.

Reply:
I've added a relatively simple cache layer.

- cache key is based on the file path and file contents
- PostCSS parsing is skipped on a cache hit

On very large prop files and codebases with many CSS files that are processed in parallel (something like a large Next JS thing) this will definitely be noticeable.

I tested with a 2mb props file and 10 CSS source files in parallel.
With the plugin on the main branch this took ±1.2s on an M2 mac.
After the changes it was ±0.35s and the number of files only has a small effect.

IO speeds on these new macs are fairly high, so results will differ.

Caching with the file path and last modified as the cache key is also an option, but might result in some false positives.

----

I've also added some tests to illustrate the failures that can happen when a tool processes files in parallel.